File: pages/CreateInboundProductPage.js
const { By, until, Key } = require("selenium-webdriver");

class CreateInboundProductPage {
  constructor(driver) {
    this.driver = driver;
  }
  dropdownInboundBtn = By.xpath("//button[contains(.,'Tạo phiếu')]");
  createInboundMenuItem = By.xpath("//button[contains(.,'Tạo phiếu nhập')]");

  warehouseField = By.xpath(
    "//div[text()='Chọn địa chỉ lấy hàng']/ancestor::div[contains(@class,'-control')]",
  );

  warehouseInput = By.xpath(
    "//div[text()='Chọn địa chỉ lấy hàng']/following::input[1]",
  );

  warehouseOption = By.xpath(
    "//*[contains(@class,'-menu')]//*[contains(normalize-space(.),'PK100270')]",
  );

  supplierField = By.xpath(
    "//div[text()='Chọn nhà cung cấp']/ancestor::div[contains(@class,'-control')]",
  );

  supplierInput = By.xpath(
    "//div[text()='Chọn nhà cung cấp']/following::input[1]",
  );

  supplierOption = By.xpath(
    "//*[contains(@class,'-menu')]//*[contains(normalize-space(.),'Supplier A')]",
  );

  referenceField = By.xpath("//input[@name='shipmentReferenceCode']");

  addProductBtn = By.xpath("//p[text()='Thêm sản phẩm']");

  addProductField = By.xpath(
    "//div[text()='Chọn sản phẩm']/ancestor::div[contains(@class,'-control')]",
  );

  addNewProductBtn = By.xpath("//button[text()='Thêm sản phẩm mới']");

  productDropdowns = By.xpath(
    "//div[text()='Chọn sản phẩm']/ancestor::div[contains(@class,'-control')]",
  );

  productOption = (keyword) =>
    By.xpath(
      `//*[contains(@class,'-menu')]//*[contains(normalize-space(.),'${keyword}')]`,
    );

  productQtyInputByRow = (rowIndex) =>
    By.xpath(`//input[@name='listProduct.${rowIndex - 1}.productQty']`);

  confirmItemInboundBtn = By.xpath("//button[text()='Xác nhận']");

  lengthField = By.xpath("//input[@placeholder='Dài']");
  widthField = By.xpath("//input[@placeholder='Rộng']");
  heightField = By.xpath("//input[@placeholder='Cao']");

  createInboundBtn = By.xpath("//button[text()='Tạo mới']");
  confirmInboundBtn = By.xpath("//button[text()='Tạo và duyệt phiếu nhập']");

  inboundCodeText = By.xpath("//h5[contains(.,'Mã nhập kho')]");

  async clickCreateInboundBtn() {
    try {
      const dropdownBtn = await this.driver.wait(
        until.elementLocated(this.dropdownInboundBtn),
        5000,
      );
      await dropdownBtn.click();
      console.log("Clicked 'Tạo phiếu' button");
    } catch (error) {
      console.error(
        "Không tìm thấy nút 'Tạo phiếu' hoặc có lỗi khi click: ",
        error,
      );
    }
  }
  async clickCreateInboundMenuItem() {
    try {
      const menuItem = await this.driver.wait(
        until.elementLocated(this.createInboundMenuItem),
        5000,
      );
      await menuItem.click();
      console.log("Clicked 'Tạo phiếu nhập' menu item");
    } catch (error) {
      console.error(
        "Không tìm thấy menu item 'Tạo phiếu nhập' hoặc có lỗi khi click: ",
        error,
      );
    }
  }
  async selectWarehouse() {
    // mở dropdown
    const dropdown = await this.driver.wait(
      until.elementLocated(this.warehouseField),
      10000,
    );

    await this.driver.executeScript("arguments[0].click();", dropdown);

    console.log("Opened warehouse dropdown");

    // input search
    const input = await this.driver.wait(
      until.elementLocated(this.warehouseInput),
      10000,
    );

    // nhập keyword
    await input.sendKeys("PK100270");

    console.log("Typed warehouse keyword");

    // wait option render
    const option = await this.driver.wait(
      until.elementLocated(this.warehouseOption),
      10000,
    );

    await this.driver.wait(until.elementIsVisible(option), 10000);

    console.log("Warehouse option appeared");

    // press ENTER để select
    await input.sendKeys(Key.ENTER);

    console.log("Selected warehouse by ENTER");
  }

  async selectSupplier() {
    // mở dropdown
    const dropdown = await this.driver.wait(
      until.elementLocated(this.supplierField),
      10000,
    );
    await this.driver.executeScript("arguments[0].click();", dropdown);
    console.log("Opened supplier dropdown");

    // input search
    const input = await this.driver.wait(
      until.elementLocated(this.supplierInput),
      10000,
    );
    // nhập keyword
    await input.sendKeys("Supplier A");
    console.log("Typed supplier keyword");
    // wait option render
    const option = await this.driver.wait(
      until.elementLocated(this.supplierOption),
      10000,
    );
    await this.driver.wait(until.elementIsVisible(option), 10000);
    console.log("Supplier option appeared");
    // press ENTER để select
    await input.sendKeys(Key.ENTER);
    console.log("Selected supplier by ENTER");
  }

  async inputReference() {
    const randomNUM = Date.now().toString().slice(-6);
    const reference = `REF-${randomNUM}`;
    const referenceInput = await this.driver.wait(
      until.elementLocated(this.referenceField),
      10000,
    );
    await referenceInput.sendKeys(reference);
    console.log("Entered reference: " + reference);
  }

  async clickAddProductBtn() {
    const addProductBtn = await this.driver.wait(
      until.elementLocated(this.addProductBtn),
      10000,
    );
    await addProductBtn.click();
    console.log("Clicked 'Thêm sản phẩm' button");
  }

  async getLastProductDropdown() {
    const dropdowns = await this.driver.wait(async () => {
      const elements = await this.driver.findElements(this.productDropdowns);
      return elements.length > 0 ? elements : false;
    }, 10000);

    return dropdowns[dropdowns.length - 1];
  }

  async addProductToInbound(productKeyword, qty = 1) {
    const dropdown = await this.getLastProductDropdown();

    await this.driver.executeScript(
      "arguments[0].scrollIntoView({block:'center'});",
      dropdown,
    );

    await this.driver.executeScript("arguments[0].click();", dropdown);

    const input = await dropdown.findElement(By.css("input"));

    await input.sendKeys(productKeyword);

    const option = await this.driver.wait(
      until.elementLocated(this.productOption(productKeyword)),
      10000,
    );

    await this.driver.wait(until.elementIsVisible(option), 10000);

    await input.sendKeys(Key.ENTER);

    // Sau khi chọn product, nhập số lượng cho dòng cuối cùng
    const qtyInputs = await this.driver.findElements(
      By.xpath("//input[contains(@name,'productQty')]"),
    );

    const qtyInput = qtyInputs[qtyInputs.length - 1];

    await qtyInput.clear();
    await qtyInput.sendKeys(String(qty));

    console.log(`Selected product ${productKeyword} with qty ${qty}`);
  }

  async clickAddNewProductBtn() {
    const beforeRows = await this.driver.findElements(this.productDropdowns);
    const beforeCount = beforeRows.length;

    const btn = await this.driver.wait(
      until.elementLocated(this.addNewProductBtn),
      10000,
    );

    await btn.click();

    await this.driver.wait(async () => {
      const rows = await this.driver.findElements(this.productDropdowns);
      return rows.length > beforeCount;
    }, 10000);

    console.log("Added new product row");
  }

  async clickItemsConfirmInbound() {
    const btn = await this.driver.wait(
      until.elementLocated(this.confirmItemInboundBtn),
      10000,
    );
    await btn.click();
    console.log("Clicked 'Xác nhận' button to create inbound");
  }

  async inputProductDimensions(length, width, height) {
    const lengthInput = await this.driver.wait(
      until.elementLocated(this.lengthField),
      10000,
    );
    await lengthInput.clear();
    await lengthInput.sendKeys(String(length));
    const widthInput = await this.driver.wait(
      until.elementLocated(this.widthField),
      10000,
    );
    await widthInput.clear();
    await widthInput.sendKeys(String(width));

    const heightInput = await this.driver.wait(
      until.elementLocated(this.heightField),
      10000,
    );
    await heightInput.clear();
    await heightInput.sendKeys(String(height));
    console.log(`Entered product dimensions: ${length}x${width}x${height}`);
  }

  async confirmCreateInbound() {
    const createBtn = await this.driver.wait(
      until.elementLocated(this.createInboundBtn),
      10000,
    );
    await createBtn.click();
    console.log("Clicked 'Tạo mới' button to create inbound");

    const createAndApprovedbtn = await this.driver.wait(
      until.elementLocated(this.confirmInboundBtn),
      10000,
    );
    await createAndApprovedbtn.click();
    console.log(
      "Clicked 'Tạo và duyệt phiếu nhập' button to confirm inbound creation",
    );
  }

  async getInboundCode() {
    const inboundElement = await this.driver.wait(
      until.elementLocated(this.inboundCodeText),
      10000,
    );

    const text = await inboundElement.getText();

    console.log("Inbound text:", text);

    // regex lấy mã NHIV...
    const match = text.match(/NHIV\d+/);

    if (!match) {
      throw new Error("Không tìm thấy mã nhập kho");
    }

    const inboundCode = match[0];

    console.log("Inbound Code:", inboundCode);

    return inboundCode;
  }
}
module.exports = CreateInboundProductPage;

File: src/test/java/com/example/seleniumtestng/pages/CreateInboundProductPage.java
package com.example.seleniumtestng.pages;

import java.util.List;
import java.util.stream.Collectors;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import org.openqa.selenium.By;
import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.Keys;
import org.openqa.selenium.TimeoutException;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;

public class CreateInboundProductPage extends BasePage {
    private final By dropdownInboundBtn = By.xpath("//button[contains(.,'Tạo phiếu') or contains(.,'Tao phieu')]");
    private final By createInboundMenuItem = By.xpath("//button[contains(.,'Tạo phiếu nhập') or contains(.,'Tao phieu nhap')]");
    private final By warehouseField = By.xpath("//div[contains(.,'Chọn địa chỉ lấy hàng') or contains(.,'Chon dia chi lay hang')]/ancestor::div[contains(@class,'-control')]");
    private final By warehouseInput = By.xpath("//div[contains(.,'Chọn địa chỉ lấy hàng') or contains(.,'Chon dia chi lay hang')]/following::input[1]");
    private final By supplierField = By.xpath("//div[contains(.,'Chọn nhà cung cấp') or contains(.,'Chon nha cung cap')]/ancestor::div[contains(@class,'-control')]");
    private final By supplierInput = By.xpath("//div[contains(.,'Chọn nhà cung cấp') or contains(.,'Chon nha cung cap')]/following::input[1]");
    private final By reactSelectControls = By.cssSelector("div[class*='-control']");
    private final By referenceField = By.cssSelector("input[name='shipmentReferenceCode']");
    private final By addProductBtn = By.xpath("//*[normalize-space()='Thêm sản phẩm' or normalize-space()='Them san pham']");
    private final By addNewProductBtn = By.xpath("//button[normalize-space()='Thêm sản phẩm mới' or normalize-space()='Them san pham moi']");
    private final By productDropdowns = By.xpath("//div[contains(.,'Chọn sản phẩm') or contains(.,'Chon san pham')]/ancestor::div[contains(@class,'-control')]");
    private final By confirmItemInboundBtn = By.xpath("//button[normalize-space()='Xác nhận' or normalize-space()='Xac nhan']");
    private final By lengthField = By.xpath("//input[contains(@placeholder,'Dài') or contains(@placeholder,'Dai')]");
    private final By widthField = By.xpath("//input[contains(@placeholder,'Rộng') or contains(@placeholder,'Rong')]");
    private final By heightField = By.xpath("//input[contains(@placeholder,'Cao')]");
    private final By createInboundBtn = By.xpath("//button[normalize-space()='Tạo mới' or normalize-space()='Tao moi']");
    private final By confirmInboundBtn = By.xpath("//button[normalize-space()='Tạo và duyệt phiếu nhập' or normalize-space()='Tao va duyet phieu nhap']");
    private final By inboundCodeText = By.xpath("//h5[contains(.,'Mã nhập kho') or contains(.,'Ma nhap kho')]");

    public CreateInboundProductPage(WebDriver driver) {
        super(driver);
    }

    public void openCreateInboundForm() {
        click(dropdownInboundBtn);
        click(createInboundMenuItem);
    }

    public void selectWarehouse(String keyword) {
        selectReactOption(warehouseField, warehouseInput, keyword, 0);
    }

    public void selectSupplier(String keyword) {
        selectReactOption(supplierField, supplierInput, keyword, 1);
    }

    public String inputReference() {
        String reference = "REF-" + String.valueOf(System.currentTimeMillis()).substring(7);
        type(referenceField, reference);
        return reference;
    }

    public boolean continueIfPresent() {
        return clickTextActionIfPresent("tiep tuc", 0, false, 5000);
    }

    public void clickAddProduct() {
        click(addProductBtn);
    }

    public void clickAddProductForPackage(int packageIndex) {
        clickTextAction("them san pham", packageIndex, true);
    }

    public void addProductToInbound(String productKeyword, int quantity) {
        WebElement dropdown = last(productDropdowns);
        jsClick(dropdown);
        WebElement input = dropdown.findElement(By.cssSelector("input"));
        input.sendKeys(productKeyword);
        visible(By.xpath("//*[contains(@class,'-menu')]//*[contains(normalize-space(.)," + xpathText(productKeyword) + ")]"));
        input.sendKeys(Keys.ENTER);

        List<WebElement> qtyInputs = all(By.xpath("//input[contains(@name,'productQty')]"));
        WebElement qtyInput = qtyInputs.get(qtyInputs.size() - 1);
        qtyInput.clear();
        qtyInput.sendKeys(String.valueOf(quantity));
    }

    public void addNewProductRow() {
        int before = all(productDropdowns).size();
        click(addNewProductBtn);
        wait.until(driver -> all(productDropdowns).size() > before);
    }

    public void addPackageRow() {
        clickTextAction("them kien", 0, false);
    }

    public void confirmItems() {
        click(confirmItemInboundBtn);
    }

    public void inputProductDimensions(int length, int width, int height) {
        type(lengthField, String.valueOf(length));
        type(widthField, String.valueOf(width));
        type(heightField, String.valueOf(height));
    }

    public void inputPackageDimensions(int packageIndex, int length, int width, int height) {
        type(indexedVisible(lengthField, packageIndex), String.valueOf(length));
        type(indexedVisible(widthField, packageIndex), String.valueOf(width));
        type(indexedVisible(heightField, packageIndex), String.valueOf(height));
    }

    public void confirmCreateInbound() {
        clickCreateInboundDropdown();
        if (!clickTextActionIfPresent("tao va duyet phieu nhap", 0, false, 5000)) {
            clickCreateInboundDropdown();
            if (!clickTextActionIfPresent("tao va duyet phieu nhap", 0, false, 5000)) {
                throw new TimeoutException("Create and approve inbound option not found. Buttons=" + buttonSummary());
            }
        }
        clickTextActionIfPresent("xac nhan", 0, false, 3000);
    }

    public String getInboundCode() {
        String inboundCode = shortWait(30000).until(driver -> {
            String fromUrl = firstInboundCode(driver.getCurrentUrl());
            if (fromUrl != null) {
                return fromUrl;
            }
            for (WebElement element : all(inboundCodeText)) {
                if (!displayed(element)) {
                    continue;
                }
                String fromLabel = firstInboundCode(element.getText());
                if (fromLabel != null) {
                    return fromLabel;
                }
            }
            String fromBody = firstInboundCode(driver.findElement(By.tagName("body")).getText());
            return fromBody == null ? null : fromBody;
        });
        System.out.println("Created inbound PO: " + inboundCode);
        return inboundCode;
    }

    private String firstInboundCode(String value) {
        if (value == null || value.isBlank()) {
            return null;
        }
        Matcher matcher = Pattern.compile("NHIV\\d+").matcher(value);
        return matcher.find() ? matcher.group() : null;
    }

    private void clickCreateInboundDropdown() {
        for (int attempt = 0; attempt < 3; attempt++) {
            if (clickButtonByNormalizedTextIfPresent("tao moi", 3000)
                    || clickButtonByNormalizedTextIfPresent("tao phieu nhap", 3000)) {
                return;
            }
            if (!clickTextActionIfPresent("tiep tuc", 0, false, 5000)) {
                break;
            }
        }
        throw new TimeoutException("Create inbound action button not found. Buttons=" + buttonSummary());
    }

    private boolean clickButtonByNormalizedTextIfPresent(String normalizedText, long timeoutMillis) {
        try {
            WebElement button = shortWait(timeoutMillis).until(driver -> {
                for (WebElement candidate : all(By.cssSelector("button"))) {
                    if (!displayed(candidate) || !candidate.isEnabled()) {
                        continue;
                    }
                    if (normalizeForMatch(candidate.getText()).contains(normalizedText)) {
                        return candidate;
                    }
                }
                return null;
            });
            try {
                button.click();
            } catch (RuntimeException e) {
                jsClick(button);
            }
            return true;
        } catch (TimeoutException e) {
            return false;
        }
    }

    private void selectReactOption(By field, By inputLocator, String keyword, int fallbackControlIndex) {
        WebElement dropdown = reactSelectControl(field, fallbackControlIndex);
        try {
            dropdown.click();
        } catch (RuntimeException e) {
            jsClick(dropdown);
        }
        WebElement input = inputInside(dropdown, inputLocator);
        input.sendKeys(keyword);
        visible(By.xpath("//*[contains(@class,'-menu')]//*[contains(normalize-space(.)," + xpathText(keyword) + ")]"));
        input.sendKeys(Keys.ENTER);
    }

    private WebElement reactSelectControl(By preferredLocator, int fallbackControlIndex) {
        try {
            return shortWait(3000).until(driver -> {
                List<WebElement> preferred = visibleElements(preferredLocator);
                return preferred.isEmpty() ? null : preferred.get(0);
            });
        } catch (TimeoutException ignored) {
            return indexedVisible(reactSelectControls, fallbackControlIndex);
        }
    }

    private WebElement inputInside(WebElement dropdown, By fallbackInputLocator) {
        try {
            return dropdown.findElement(By.cssSelector("input"));
        } catch (RuntimeException ignored) {
            return visible(fallbackInputLocator);
        }
    }

    private WebElement last(By locator) {
        return wait.until(driver -> {
            List<WebElement> elements = all(locator);
            return elements.isEmpty() ? null : elements.get(elements.size() - 1);
        });
    }

    private WebElement indexedVisible(By locator, int index) {
        return wait.until(driver -> {
            List<WebElement> elements = visibleElements(locator);
            return elements.size() > index ? elements.get(index) : null;
        });
    }

    private void clickTextAction(String normalizedText, int index, boolean excludeAddNewProduct) {
        if (clickTextActionIfPresent(normalizedText, index, excludeAddNewProduct, 15000)) {
            return;
        }
        throw new TimeoutException("Button not found: text="
                + normalizedText
                + ", index="
                + index
                + ", buttons="
                + buttonSummary());
    }

    private boolean clickTextActionIfPresent(
            String normalizedText,
            int index,
            boolean excludeAddNewProduct,
            long timeoutMillis) {
        try {
            shortWait(timeoutMillis).until(driver -> {
                Object clicked = ((JavascriptExecutor) driver).executeScript(
                        "const target = arguments[0];"
                                + "const index = arguments[1];"
                                + "const excludeAddNew = arguments[2];"
                                + "const norm = value => (value || '')"
                                + "  .normalize('NFD').replace(/[\\u0300-\\u036f]/g, '')"
                                + "  .replace(/đ/g, 'd').replace(/Đ/g, 'D')"
                                + "  .toLowerCase().replace(/\\s+/g, ' ').trim();"
                                + "const visible = el => {"
                                + "  const style = window.getComputedStyle(el);"
                                + "  const rect = el.getBoundingClientRect();"
                                + "  return style.visibility !== 'hidden' && style.display !== 'none'"
                                + "    && rect.width > 0 && rect.height > 0;"
                                + "};"
                                + "const matches = el => {"
                                + "  const text = norm(el.innerText || el.textContent);"
                                + "  return text.includes(target) && (!excludeAddNew || !text.includes('them san pham moi'));"
                                + "};"
                                + "const selector = '*';"
                                + "const elements = Array.from(document.querySelectorAll(selector))"
                                + "  .filter(el => visible(el) && matches(el))"
                                + "  .filter(el => !Array.from(el.children).some(child => visible(child) && matches(child)));"
                                + "if (!elements.length) return false;"
                                + "const element = elements[Math.min(index, elements.length - 1)];"
                                + "const clickable = element.closest('button,a,[role=\"button\"],[onclick],.btn,[class*=\"button\"],[class*=\"cursor\"]') || element;"
                                + "clickable.scrollIntoView({block:'center'});"
                                + "clickable.click();"
                                + "return true;",
                        normalizedText,
                        index,
                        excludeAddNewProduct);
                return Boolean.TRUE.equals(clicked) ? true : null;
            });
            return true;
        } catch (TimeoutException e) {
            return false;
        }
    }

    private WebElement indexedVisibleButton(String normalizedText, int index) {
        try {
            return wait.until(driver -> {
                List<WebElement> elements = visibleTextElements(normalizedText, false);
                if (elements.isEmpty()) {
                    return null;
                }
                return elements.get(Math.min(index, elements.size() - 1));
            });
        } catch (TimeoutException e) {
            throw new TimeoutException("Button not found: text="
                    + normalizedText
                    + ", index="
                    + index
                    + ", buttons="
                    + buttonSummary(), e);
        }
    }

    private List<WebElement> visibleTextElements(String normalizedText, boolean excludeAddNewProduct) {
        return all(By.xpath("//*[normalize-space()]")).stream()
                .filter(this::displayed)
                .filter(element -> textMatches(element, normalizedText, excludeAddNewProduct))
                .filter(element -> !hasMatchingChild(element, normalizedText, excludeAddNewProduct))
                .collect(Collectors.toList());
    }

    private boolean textMatches(WebElement element, String normalizedText, boolean excludeAddNewProduct) {
        String text = normalizeForMatch(element.getText());
        return text.contains(normalizedText)
                && (!excludeAddNewProduct || !text.contains("them san pham moi"));
    }

    private boolean hasMatchingChild(WebElement element, String normalizedText, boolean excludeAddNewProduct) {
        for (WebElement child : element.findElements(By.xpath(".//*[normalize-space()]"))) {
            if (displayed(child) && textMatches(child, normalizedText, excludeAddNewProduct)) {
                return true;
            }
        }
        return false;
    }

    private String buttonSummary() {
        return all(By.cssSelector("button")).stream()
                .filter(this::displayed)
                .map(button -> {
                    String text = button.getText().replaceAll("\\s+", " ").trim();
                    return text + " => " + normalizeForMatch(text);
                })
                .collect(Collectors.joining(" | "));
    }

    private List<WebElement> visibleElements(By locator) {
        return all(locator).stream()
                .filter(this::displayed)
                .collect(Collectors.toList());
    }

    private void type(WebElement element, String value) {
        jsClick(element);
        element.clear();
        element.sendKeys(value);
    }
}
